apptests/test_testrunners.py

```diff
@@ -0,0 +1,234 @@
+            save_to_database=False,
+            keep_audio=True,
+        )
+        mock_video = MagicMock(spec=YouTubeVideo)
+        mock_video.youtube_transcript.return_value = "This is a model for tests :)"
+        mock_video.download_mp3.return_value = "./apptests/data/sample.mp3"
+        mock_from_dict.return_value = mock_video
+
+        mock_dummy_test_instance.compare.return_value = {
+            "result": "Mock Compare Result"
+        }
+
+        read_data = '{"args": {"q": "test"}, "items": [{"videoId": "123"}], "nextPageToken": "token"}'
+
+        mock_open_func = mock_open(read_data=read_data)
+        mock_open_func.return_value.__enter__.return_value.read.return_value = read_data
+
+        def custom_json_dump(obj, f, ensure_ascii, indent):
+            serializable_obj = {}
+            for key, value in obj.items():
+                if isinstance(value, Mock):
+                    serializable_obj[
+                        key
+                    ] = f"Mock object of type {value._spec_class.__name__}"
+                else:
+                    serializable_obj[key] = value
+
+            json.dump(serializable_obj, f, ensure_ascii=ensure_ascii, indent=indent)
+
+        mock_json_dump.side_effect = custom_json_dump
+
+        with patch("builtins.open", mock_open_func), patch(
+            "json.load"
+        ) as mock_json_load:
+            mock_json_load.return_value = {
+                "args": {"q": "test"},
+                "items": [{"videoId": "123"}],
+                "nextPageToken": "token",
+                "language": "en",
+                "targetTranscript": "Mock Target Transcript",
+            }
+            runner.run()
+            mock_generate.assert_called_once_with(
+                mock_json_load.return_value,
+                mock_open_func.return_value.__enter__.return_value,
+                ensure_ascii=False,
+            )
+
+        mock_open.assert_called_once_with(read_data=read_data)
+        mock_dummy_test_instance.additional_info.assert_called_once()
+
+    @patch("src.dataclasses.youtube_video.YouTubeVideo.from_dict")
+    @patch("builtins.open")
+    def test_run_with_exceptions(self, mock_from_dict, mock_open):
+        mock_file = Mock()
+        mock_video = Mock()
+        mock_open.return_value = mock_file
+
+        mock_from_dict.return_value = mock_video
+
+        self.runner.tester.transcriber = None
+        self.runner.tester.normalizer = None
+        self.runner._iterations = 2
+
+        self.runner._testplan_path = "mock_testplan_path"
+
+        with patch("builtins.open"), patch(
+            "src.dataclasses.youtube_video.YouTubeVideo.from_dict"
+        ), patch("youtube_runner.YouTubeTestRunner.save_results"), patch(
+            "generators.youtube_generator.generate"
+        ):
+            with self.assertRaises(ValueError) as context:
+                self.runner.run()
+
+        self.assertEqual(str(context.exception), "Transcriber is None")
```

Comment:
I think some of the mock logic can be moved to the setUp method. setUp is called for each test method, so mocks can be created in this method and later slightly modified in the test method.

Reply:
i moved logic to setup